presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -714,6 +714,16 @@ protected ExtractionResult visitInPredicate(InPredicate node, Boolean complement
             return extractionResult;
         }
 
+        private boolean canProcessExpression(Expression expression)
```

Comment:
The logic here should align with the `visitComparisonExpression` (we are short-cutting delegation to that method), but 
- it's not obvious that it is aligned today (code looks very different, eg. that method delegates to `toNormalizedSimpleComparison`)
- it's not easy to maintain it aligned (the relation between this and that method is visible in this PR only)
- are you sure this is correct? note that `expr in (x)` is equivalent to `x in (expr)` and apparently `toNormalizedSimpleComparison` is flexible enough to realize that

Instead of doing this, I would propose considering a different  approach:
- if the InListExpression is small, do not change anything
- if the InListExpression is big, try-process a short prefix of the InListExpression; if that worked, process whole InListExpression

I am aware this isn't as simple as changing `if` condition, but hopefully more maintainable.

@kokosing @martint

Reply:
@findepi Another approach is instead of translating `InPredicate` to `OR` expressions and process those expression. What if we handle at `visitInPredicate ` itself. We iterate through each expression in `InListExpression`, create a `ComparisonExpression` , process it and if TupleDomain#isAll is true we can directly skip processing other expression in the expression list. What’s your opinion on this approach ?

Reply:
That beats the advantage that @findepi is suggesting. You still have to go through the entire InListExpression before you can make any decision about short circuiting

Reply:
That beats the advantage that @findepi is suggesting. You still have to go through the entire InListExpression before you can make any decision about short circuiting

Reply:
@atris Actually we don't have to iterate through the entire `InListExpression` , say if I have a `InPredicate` like this 

`date(columnA) in ( date '2018-11-10', date '2017-11-10', ...)`

When I process the first element of `InListExpression` I first convert it to a comparison criteria 

`date(columnA) = date '2018-11-10'`

Now when I try to process this `ComparisonExpression` , the `TupleDomain#isAll`  which was extracted from the `ExtractionResult` will be true (because `TupleDomain` doesn't support FunctionCall). So now I can skip the processing for other Expression in the `InExpressionList`

Reply:
Right, so that's what @findepi is suggesting as well, I believe.

I have a suspicion that we might be going down a rabbit hole here by hard wiring a bunch of cases (If InPredicate And Cast And DomainTranslator lacks support et al). What happens when we make DomainTranslator work with FunctionCall tomorrow?

I believe the fix is in a right spirit, just needs some more thinking through. It might be worthwhile to see if we can do it more elegantly. Just my 2c

Reply:
@Praveen2112 i was also thinking about converting `InListExpression` "once", incrementally.
Didn't try to put this is in code -- would that require copying logic for handling OR? Or, maybe, OR can be called incrementally, because it's already iterative?

Note: `x IN (x)` will evaluate to `TupleDomain.all()`. I think we should try not to break too early upon `x IN (x, 1, 2, 3, 4)`. (Obviously, this is a contrived example, but also a legal one)

Reply:
@findepi FWIW, I vote for reusing OR on the expression that we are looking at currently, from the InList.

Reply:
@findepi The second option that you gave. I am not in favour of copying the logic from OR handler.

Reply:
@findepi logic for handling OR has two parts 
1. Doing a column-wise union of `TupleDomain` , which is performed at `TupleDomain` 
2. Merging the remaining expression from the `ExtractionResult` (The part of the expression which has not be translated to `TupleDomain`).

For implementing this in processing `InListExpression` we require the first part (Column wise union of TupleDomain), ideally it doesn't involve the copying the entire logic of OR. 

>Note: x IN (x) will evaluate to TupleDomain.all(). I think we should try not to break too early upon x IN (x, 1, 2, 3, 4). (Obviously, this is a contrived example, but also a legal one)

For this expression what would be the issue when we break it early ?
In the existing model this expression `x  IN (x, 1, 2, 3, 4)` will evaluate to `TupleDomain.all()` right ? (Same as `x IN (y, 1, 2, 3, 4)`). When we try to combine (column wise union ) of a existing TupleDomain with a TupleDomain.all() will evaluate to `TupleDomain.all()` right ? Please correct me If I am wrong

Reply:
> For this expression what would be the issue when we break it early ?
> In the existing model this expression x IN (x, 1, 2, 3, 4) will evaluate to TupleDomain.all() right ? ....

Ignore my comment, i made a mistake. Indeed, no problem here. 

> For implementing this in processing InListExpression we require the first part (Column wise union of TupleDomain), ideally it doesn't involve the copying the entire logic of OR.

@Praveen2112 please update the code and let's iterate from there.

Reply:
That's true but what if we have a `1 IN (x, y, z)` , it might translate to `(x = 1) or (y = 1) or (z = 1)` currently `TupleDomain` doesn't support OR between various columns but in future if it supports we easily handle them in `TupleDomain` itself right ?

Reply:
I think so, however I am not sure the word `easily` here ;)

Reply:
`easily` = `TupleDomain` handles them at his part so we don't have to worry about it